lib/core/Concurrency.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include <core/Concurrency.h>
+
+#include <core/CLogger.h>
+
+#include <boost/make_unique.hpp>
+
+#include <memory>
+#include <thread>
+
+namespace ml {
+namespace core {
+namespace {
+class CExecutorHolder {
+public:
+    CExecutorHolder()
+        : m_ThreadPoolSize{0}, m_Executor{boost::make_unique<transwarp::sequential>()} {}
+
+    static CExecutorHolder makeThreadPool(std::size_t threadPoolSize) {
+        if (threadPoolSize == 0) {
+            threadPoolSize = std::thread::hardware_concurrency();
+        }
+
+        if (threadPoolSize > 0) {
+            try {
+                return CExecutorHolder{threadPoolSize};
+            } catch (const std::exception& e) {
+                LOG_ERROR(<< "Failed to create thread pool: " << e.what());
+            }
+        } else {
+            LOG_WARN(<< "Failed to determine hardware concurrency");
```

Comment:
the choice of log level is not clear to me, why isn't it error in both cases?

I would also add the consequence in the log message: "Failed to ... , fallback to ..."

Reply:
I originally had them as both warn (because of the fallback) and I think I agree they probably should be the same. ++ on improving the message.